### ISSUE TITLE
Clang c++ compiles error with struct in template class and < operator

### ISSUE THREAD

**Issue by wangbo15** [llvm/llvm-project#81731]
Please consider the following code:

```c++
template<class T>
class confused{};

template<class T>
class test
{
    void foo() {
        if(this->b.confused < 1);
    };

    struct bar
    {
        int confused;
    } b;
};
```
Clang rejects it by giving such considerably puzzling output:

```c++
<source>:8:32: error: expected '>'
    8 |         if(this->b.confused < 1);
      |                                ^
<source>:8:29: note: to match this '<'
    8 |         if(this->b.confused < 1);
      |                             ^
<source>:8:32: error: expected unqualified-id
    8 |         if(this->b.confused < 1);
      |                                ^
<source>:8:33: warning: if statement has empty body [-Wempty-body]
    8 |         if(this->b.confused < 1);
      |                                 ^
<source>:8:33: note: put the semicolon on a separate line to silence this warning
1 warning and 2 errors generated.
Compiler returned: 1
```

We suppose that's because the compiler has found incorrect local variable name of `confused` by confusing the global template class declared above with the integer member variable declared in the nested class `bar` inside `test`

Please check [https://godbolt.org/z/7afqsvEvo](https://godbolt.org/z/7afqsvEvo).

**Comment by shafik** [comment]
This goes all the way back to clang-3.0: https://godbolt.org/z/srY7rsW94

I am surprised we have not seen this before. Could be related to: https://github.com/llvm/llvm-project/issues/78213

**Comment by shafik** [comment]
So quick results from debugging, in `Sema::isTemplateName(...)` in the failing case if we examine `ObjectType` we see:

```console
expr ObjectType->dump()

BuiltinType 0x14403de90 '<dependent type>' dependent
```

but if we remove template class `confused` we have the following:

```console
expr ObjectType->dump()

ElaboratedType 0x137008430 'struct bar' sugar dependent
`-RecordType 0x137008290 'struct test::bar' dependent
  `-CXXRecord 0x137008200 'bar'
```

I think back in `Parser::ParsePostfixExpressionSuffix(...)` I need to see what the following code is acting differently for these two cases:

```cpp
LHS = Actions.ActOnStartCXXMemberReference(getCurScope(), Base, OpLoc,
                                                   OpKind, ObjectType,
                                                   MayBePseudoDestructor);
```

If we examine `LHS` we also see differences, for the broken case:

```console
CXXDependentScopeMemberExpr 0x138145af8 '<dependent type>' lvalue ->b
`-CXXThisExpr 0x138145ae8 'test<T> *' this
```

and if we remove the template class `confused`:

```console
expr LHS.get()->dump()

MemberExpr 0x1370085c8 'struct bar':'struct test::bar' lvalue ->b 0x137008488
`-CXXThisExpr 0x1370085b8 'test<T> *' implicit this
```

**Comment by shafik** [comment]
If you look at a bunch of the examples in [basic.lookup.qual.general](https://eel.is/c++draft/basic.lookup.qual#general-3) there does not seem to be a nice simple heuristic to determining if I should consider the name found valid or not.

**Comment by shafik** [comment]
I attempted to do extra filtering in `Sema::isTemplateName(...)`

```cpp
if (R.empty())
    return TNK_Non_template;
  else if (!ObjectType.isNull() && R.isSingleResult())
   if (CXXRecordDecl *RD = ObjectType->getAsCXXRecordDecl()) {
    if (NamedDecl *ND = R.getFoundDecl()) {
      if (ClassTemplateDecl *CTD = dyn_cast<ClassTemplateDecl>(ND))
        if (CXXRecordDecl *RDF = dyn_cast<CXXRecordDecl>(CTD->getTemplatedDecl())) {
          bool AnyEqual = false;
          if (RD->getMostRecentDecl() == RDF->getMostRecentDecl())
            AnyEqual = true;

          for ( auto base : RD->getMostRecentDecl()->bases() ) {
            if (!base.getType().isNull() &&
                base.getType()->getAsCXXRecordDecl() &&
                base.getType()->getAsCXXRecordDecl()->getMostRecentDecl() == RDF->getMostRecentDecl())
              AnyEqual = true;
            
            if (!base.getType().isNull())
              if (const TemplateSpecializationType *TST = base.getType()->getAs<TemplateSpecializationType>()) {
                TemplateName TN = TST->getTemplateName();
                if (ClassTemplateDecl *CTD = dyn_cast<ClassTemplateDecl>(TN.getAsTemplateDecl()))
                  if (CXXRecordDecl *RDB = dyn_cast<CXXRecordDecl>(CTD->getTemplatedDecl()))
                    if (RDF->getMostRecentDecl() == RDB->getMostRecentDecl())
                      AnyEqual = true;
              }
          }

          if (!AnyEqual)
            return TNK_Non_template;
        }
    }
  }
```

It kind of works but I end up w/ a lot of duplicate diagnostics in several tests. It definitely feels hacky but there wasn't a more obvious place to do filtering.

**Comment by zygoloid** [comment]
Clang is interpreting the `confused<` as starting a nested name specifier, as in:

```c++
template<class T>
class confused{ int x; };

template<class T>
class test
{
    void foo() {
        if(this->b.confused<int>::x);
    };

    struct bar
    {
        int confused;
    } b;
};

template<> class test<int>::bar : confused<int> { };
// now test<int>::foo() accesses the confused<int>::x member in b
```

The rules here were changed by [P1787R6](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p1787r6.html). It used to be that if a member lookup was dependent and the name was followed by `<`, we were supposed to perform unqualified lookup to see if the name is a template or not. But after P1787, the rules are simpler: if the qualified member lookup is dependent, [we don't perform unqualified lookup at all](https://eel.is/c++draft/basic.lookup#qual.general-3.sentence-3), and then [we never treat the `<` as starting a template argument list](https://eel.is/c++draft/temp.names#3). If you mean to name a template here, you need to use the `template` keyword.

It looks like Clang doesn't implement this part of P1787R6 yet.

**Comment by zygoloid** [comment]
> If you look at a bunch of the examples in [basic.lookup.qual.general](https://eel.is/c++draft/basic.lookup.qual#general-3) there does not seem to be a nice simple heuristic to determining if I should consider the name found valid or not.

I'll try to give a bit more context about what's happening in each case:

> ```c++
>   p->X<0>::f();                 // error: A​::​X not found in ((p->X) < 0) > ​::​f()
> ```

Here, the lookup for `X` is dependent, so it's treated as a non-template, and the `<` is a less-than. Instantiation will fail with `T` = `A` because there's no member `X` in `A`.

> ```c++
>   p->template X<0>::f();        // OK, ​::​X found in definition context
> ```

Here, the `template` keyword says to treat the name `X` as a template name. The actual meaning of `X` is not determined until instantiation, but we should perform an unqualified lookup during template parsing ("in the definition context") so we have a fallback to use if there's no member named `X` within `T` during instantiation.

If there were a member template named `X` in `A`, this would name that member template instead. See the `D<0>` case below for an example of that.

> ```c++
>   p->B::f();                    // OK, non-type A​::​B ignored
> ```

Right, [this lookup for `B` ignores results that can't be used as nested-name-specifiers](https://eel.is/c++draft/basic.lookup.qual.general#1.sentence-1).

> ```c++
>   p->template C<0>::f();        // error: A​::​C is not a template
> ```

This is fine when parsing the template, and parses as a template-id not as a less-than comparison, but fails during instantiation because `C` resolves to `A::C` which is not a template but is preceded by the `template` keyword.

> ```c++
>   p->template D<0>::f();        // error: A​::​D<0> is not a class type
> ```

Like the previous example, this is fine when parsing the template, but fails in instantiation because `A::D<0>` is `void`, which can't be used as a nested name specifier. This `D` resolves to the member of the class found in the instantiation context, because notionally unqualified lookup is not performed here. (But we still need to perform unqualified lookup in the template definition context so we have a lookup result for `D` to fall back onto if it turns out that there is no member of `A` named `D`.)

> ```c++
>   p->T::f();                    // error: A​::​T is not a class type
> ```

This names `A::T` (found in the instantiation context), not the template parameter `T` (found in the definition context). If `A` had no member named `T`, this would instead name the template parameter `T` during the instantiation with `T` = `A`.

**Comment by zwuis** [comment]
Maybe related: #59910.
